pkg/cmd/server/kubernetes/node.go

```diff
@@ -194,6 +198,52 @@ func (c *NodeConfig) initializeVolumeDir(ce commandExecutor, path string) (strin
 	return rootDirectory, nil
 }
 
+// EnsureLocalQuota checks if the node config specifies a local storage
+// perFSGroup quota, and if so will test that the volumeDirectory is on a
+// filesystem suitable for quota enforcement. If checks pass the k8s emptyDir
+// volume plugin will be replaced with a wrapper version which adds quota
+// functionality.
+func (c *NodeConfig) EnsureLocalQuota(nodeConfig configapi.NodeConfig) {
+	if nodeConfig.VolumeConfig.LocalQuota.PerFSGroup == nil {
+		return
+	}
+	glog.V(4).Info("Replacing empty-dir volume plugin with quota wrapper")
+	wrappedEmptyDirPlugin := false
+
+	quotaApplicator, err := emptydir.NewQuotaApplicator(nodeConfig.VolumeDirectory)
+	if err != nil {
+		glog.Fatalf("Could not set up local quota, %s", err)
+	}
+
+	// Create a volume spec with emptyDir we can use to search for the
+	// emptyDir plugin with CanSupport:
+	emptyDirSpec := &volume.Spec{
+		Volume: &kapi.Volume{
+			VolumeSource: kapi.VolumeSource{
+				EmptyDir: &kapi.EmptyDirVolumeSource{},
+			},
+		},
+	}
+
+	for idx, plugin := range c.KubeletConfig.VolumePlugins {
+		// Can't really do type checking or use a constant here as they are not exported:
+		if plugin.CanSupport(emptyDirSpec) {
+			wrapper := emptydir.EmptyDirQuotaPlugin{
+				Wrapped:         plugin,
+				Quota:           *nodeConfig.VolumeConfig.LocalQuota.PerFSGroup,
+				QuotaApplicator: quotaApplicator,
+			}
+			c.KubeletConfig.VolumePlugins[idx] = &wrapper
+			wrappedEmptyDirPlugin = true
+		}
+	}
+	// Because we can't look for the k8s emptyDir plugin by any means that would
+	// survive a refactor, error out if we couldn't find it:
+	if !wrappedEmptyDirPlugin {
+		glog.Fatal(errors.New("unable to wrap emptyDir volume plugin for quota support"))
```

Comment:
Since this is a fatal, make it a bit more human readable:

    glog.Fatal("No plugin handling EmptyDir was found, unable to apply local quotas")